## Mean-Variance Optimisation

In [5]:
import pandas as pd
import numpy as np

In [23]:
df3 = pd.read_excel("E://FIMN36700/multi_asset_etf_data.xlsx", sheet_name='excess returns')
df3

Date       BWX       DBC       EEM       EFA       HYG       IEF  \
0   2011-02-28  0.006527  0.041620 -0.000527  0.035408  0.014264 -0.002174   
1   2011-03-31  0.008255  0.026409  0.062723 -0.024055  0.000251 -0.001719   
2   2011-04-30  0.048261  0.045015  0.026783  0.055715  0.015432  0.017868   
3   2011-05-31 -0.013447 -0.051625 -0.029301 -0.021962  0.001432  0.025156   
4   2011-06-30 -0.000270 -0.042817 -0.009528 -0.012280 -0.005878 -0.005217   
..         ...       ...       ...       ...       ...       ...       ...   
159 2024-05-31  0.009647 -0.007581  0.014935  0.046020  0.011758  0.013420   
160 2024-06-30 -0.013674 -0.005917  0.021982 -0.022466  0.000595  0.007972   
161 2024-07-31  0.033913 -0.032580  0.003854  0.021317  0.018924  0.024373   
162 2024-08-31  0.025539 -0.025794  0.004799  0.027624  0.010495  0.008478   
163 2024-09-30  0.023280  0.004912  0.070120  0.006742  0.013145  0.012029   

          IYR       PSP       QAI       SPY       TIP  
0    0.045113  0.040056  0.002493  0.034647  0.007118  
1   -0.011107  0.016495  0.005348 -0.000052  0.011730  
2    0.046089  0.058127  0.018490  0.028418  0.023236  
3    0.010233 -0.041465  0.000099 -0.011115  0.002758  
4   -0.031233 -0.042754 -0.010948 -0.017042  0.007377  
..        ...       ...       ...       ...       ...  
159  0.044692  0.034457  0.008852  0.045998  0.013189  
160  0.014565 -0.021305 -0.000965  0.031082  0.003417  
161  0.071649  0.063989  0.006681  0.007510  0.012676  
162  0.049029 -0.003754  0.002669  0.018386  0.003011  
163  0.018761  0.051810  0.010278  0.012344  0.011034  

[164 rows x 12 columns]

In [24]:
def calculate_summary_statistics(df, annual_factor):
    summary_statistics = pd.DataFrame(index=df.columns)
    summary_statistics['Mean'] = df.mean(axis=0) * annual_factor
    summary_statistics['Vol'] = df.std(axis=0) * np.sqrt(annual_factor)
    summary_statistics['Sharpe'] = (df.mean() / df.std()) * np.sqrt(annual_factor)
    summary_statistics['Min'] = df.min()
    summary_statistics['Max'] = df.max()
    summary_statistics['Skewness'] = df.skew()
    summary_statistics['Excess Kurtosis'] = df.kurtosis()
    summary_statistics['VaR (0.05)'] = df.quantile(.05, axis = 0)
    summary_statistics['CVaR (0.05)'] = df[df <= df.quantile(.05, axis = 0)].mean()
    return summary_statistics

In [25]:
df1 = df3

In [26]:
df3 = df3.drop(columns='Date')
summary = calculate_summary_statistics(df3, annual_factor=12)
summary

Mean       Vol    Sharpe       Min       Max  Skewness  \
BWX -0.011888  0.081671 -0.145563 -0.069406  0.077951 -0.184804   
DBC -0.009086  0.168455 -0.053935 -0.177563  0.102184 -0.491710   
EEM  0.026960  0.179940  0.149829 -0.179084  0.162806 -0.104661   
EFA  0.055037  0.152203  0.361605 -0.145228  0.142902 -0.266527   
HYG  0.037356  0.077289  0.483335 -0.104440  0.085044 -0.482790   
IEF  0.013939  0.063197  0.220561 -0.047632  0.046105 -0.029644   
IYR  0.077912  0.169585  0.459426 -0.200485  0.132024 -0.458684   
PSP  0.092851  0.215238  0.431386 -0.278880  0.178449 -0.641633   
QAI  0.014959  0.049007  0.305241 -0.060508  0.036397 -0.538754   
SPY  0.126983  0.143066  0.887578 -0.129032  0.126767 -0.407580   
TIP  0.016844  0.051258  0.328618 -0.067605  0.042217 -0.677103   

     Excess Kurtosis  VaR (0.05)  CVaR (0.05)  
BWX         0.714401   -0.043311    -0.052614  
DBC         0.708068   -0.083417    -0.113114  
EEM         1.277024   -0.078424    -0.111063  
EFA         0.967185   -0.073256    -0.096643  
HYG         4.299609   -0.027397    -0.049028  
IEF         0.098295   -0.030108    -0.038691  
IYR         1.478125   -0.067934    -0.100602  
PSP         2.574721   -0.100326    -0.140010  
QAI         2.018508   -0.019026    -0.031692  
SPY         0.858827   -0.065493    -0.085439  
TIP         2.736283   -0.021407    -0.033128

In [27]:
# print the asset with best and worst Sharpe ratio
print("The asset with the best Sharpe ratio is: ")
display(summary[summary['Sharpe'] == summary['Sharpe'].max()][['Sharpe']])
print("The asset with the worst Sharpe ratio is: ")
display(summary[summary['Sharpe'] == summary['Sharpe'].min()][['Sharpe']])

The asset with the best Sharpe ratio is: 


Sharpe
SPY  0.887578

The asset with the worst Sharpe ratio is: 


Sharpe
BWX -0.145563

2. Descriptive Analysis

In [28]:
# Calculate the correlation matrix
correlation_matrix = df3.corr()
print(correlation_matrix)
correlation_pairs = correlation_matrix.unstack()

# Exclude autocorrelation (cases where the correlation coefficient is 1)
correlation_pairs = correlation_pairs[correlation_pairs != 1]

# Identify the highest and lowest correlations along with their corresponding asset pairs
highest_corr_pair = correlation_pairs.idxmax()  
lowest_corr_pair = correlation_pairs.idxmin()   
highest_corr_value = correlation_pairs.max() 
lowest_corr_value = correlation_pairs.min()  
print(f'Highest correlation: {highest_corr_pair[0]} and {highest_corr_pair[1]} with value {highest_corr_value}')
print(f'Lowest correlation: {lowest_corr_pair[0]} and {lowest_corr_pair[1]} with value {lowest_corr_value}')

          BWX       DBC       EEM       EFA       HYG       IEF       IYR  \
BWX  1.000000  0.233222  0.638705  0.600620  0.620946  0.568903  0.562599   
DBC  0.233222  1.000000  0.516908  0.521447  0.471924 -0.300523  0.290040   
EEM  0.638705  0.516908  1.000000  0.820770  0.694174  0.020477  0.589003   
EFA  0.600620  0.521447  0.820770  1.000000  0.790742  0.018537  0.705926   
HYG  0.620946  0.471924  0.694174  0.790742  1.000000  0.180087  0.739567   
IEF  0.568903 -0.300523  0.020477  0.018537  0.180087  1.000000  0.297047   
IYR  0.562599  0.290040  0.589003  0.705926  0.739567  0.297047  1.000000   
PSP  0.552010  0.467111  0.761796  0.910596  0.814182  0.017486  0.753810   
QAI  0.662425  0.494467  0.791130  0.863957  0.811161  0.176287  0.718397   
SPY  0.466152  0.448389  0.699939  0.864822  0.794782 -0.005053  0.760646   
TIP  0.680001  0.114379  0.379465  0.386677  0.541758  0.745025  0.590805   

          PSP       QAI       SPY       TIP  
BWX  0.552010  0.662425  0.46

In [29]:
# perform the analysis on the performance of TIP
tip_row = summary.loc["TIP"]
# ranking in summary_stats
ranks = tip_row.rank(method='dense', ascending=False)
print(ranks)

Mean               5.0
Vol                3.0
Sharpe             2.0
Min                8.0
Max                4.0
Skewness           9.0
Excess Kurtosis    1.0
VaR (0.05)         6.0
CVaR (0.05)        7.0
Name: TIP, dtype: float64


According to the analysis above, TIPS exhibits relatively low annualized returns and volatility among all bonds, and its Sharpe ratio is right in the middle. It does not show a significant high correlation with other types of bonds, with the max 0.745025 with IEF and the min 0.114379 with DBC. In addition, comparing with both domestic bonds (IEF) and foreign bonds (BWX), TIP has higher mean return, lower volatility and higher sharp ratio, which means it's better than them. 

3. The MV Frontier

In [30]:
def tangency_weights(df, cov_mat = 1):
    if cov_mat ==1:
        cov_inv = np.linalg.inv((df.cov()*12))
    else:
        cov = df.cov()
        covmat_diag = np.diag(np.diag((cov)))
        covmat = cov_mat * cov + (1-cov_mat) * covmat_diag
        cov_inv = np.linalg.inv((covmat*12))  
    ones = np.ones(df.columns.shape) 
    mu = df.mean()*12
    scaling = 1/(np.transpose(ones) @ cov_inv @ mu)
    tangent_return = scaling*(cov_inv @ mu) 
    tangency_wts = pd.DataFrame(index = df.columns[:], data = tangent_return, columns = ['Tangent Weights'] )
    return tangency_wts

In [31]:
w_t = tangency_weights(df3)
mv_frontier = w_t.merge(summary['Sharpe'], on=w_t.index).rename(columns={'key_0':'assets'}).set_index('assets')
mv_frontier

Tangent Weights    Sharpe
assets                           
BWX            2.730253 -0.145563
DBC           -0.111241 -0.053935
EEM           -0.615698  0.149829
EFA            0.530692  0.361605
HYG           -0.832578  0.483335
IEF           -4.682480  0.220561
IYR            0.877108  0.459426
PSP            0.277002  0.431386
QAI            7.220592  0.305241
SPY           -4.904086  0.887578
TIP            0.510436  0.328618

In [32]:
ranked_mv_frontier = mv_frontier.rank(method='min')
print(ranked_mv_frontier)

        Tangent Weights  Sharpe
assets                         
BWX                10.0     1.0
DBC                 5.0     2.0
EEM                 4.0     3.0
EFA                 8.0     7.0
HYG                 3.0    10.0
IEF                 2.0     4.0
IYR                 9.0     9.0
PSP                 6.0     8.0
QAI                11.0     5.0
SPY                 1.0    11.0
TIP                 7.0     6.0


The ranking of weights and the ranking of Sharpe ratios are not consistent, and there are many completely opposite cases. This may indicate that the tangency portfolio we obtained is not located on the efficient mean-variance frontier. To change this situation, it is necessary to impose constraints on the minimum return.

In [33]:
w_tan_summary_statistics = calculate_summary_statistics(df3 @ w_t , annual_factor = 12)
w_tan_summary_statistics

Mean       Vol    Sharpe       Min       Max  Skewness  \
Tangent Weights -0.527276  0.328513 -1.605037 -0.293516  0.243644  -0.05178   

                 Excess Kurtosis  VaR (0.05)  CVaR (0.05)  
Tangent Weights         0.105926   -0.198018    -0.244705

4. Tips

Without TIPS

In [34]:
w_t_without_tips = tangency_weights(df3.drop(columns=['TIP']))
w_t_without_tips.columns = ['Tangency Weights with TIPS dropped']
w_t_without_tips

Tangency Weights with TIPS dropped
BWX                            3.015911
DBC                           -0.090968
EEM                           -0.664230
EFA                            0.540606
HYG                           -0.871637
IEF                           -4.832385
IYR                            0.990156
PSP                            0.324898
QAI                            7.976911
SPY                           -5.389260

In [35]:
w_tan_wo_tips_summary_statistics = calculate_summary_statistics(df3.drop(columns=['TIP']) @ w_t_without_tips , annual_factor = 12)
w_tan_wo_tips_summary_statistics.index = ['Tangency Weights with TIPS dropped']
w_tan_wo_tips_summary_statistics

Mean       Vol    Sharpe      Min  \
Tangency Weights with TIPS dropped -0.580806  0.362125 -1.603881 -0.33166   

                                         Max  Skewness  Excess Kurtosis  \
Tangency Weights with TIPS dropped  0.266782 -0.085016         0.162678   

                                    VaR (0.05)  CVaR (0.05)  
Tangency Weights with TIPS dropped   -0.220747    -0.272286

TIPS is adjusted

In [36]:
TIP_adj = 0.0012
df_adj = df3.copy()
df_adj['TIP'] += TIP_adj
w_t_adj_tips = tangency_weights(df_adj)
w_t_adj_tips.columns = ['Tangency weights with TIPS adjusted']
w_t_adj_tips

Tangency weights with TIPS adjusted
BWX                            29.260095
DBC                             1.771624
EEM                            -5.123028
EFA                             1.451443
HYG                            -4.460119
IEF                           -18.604613
IYR                            11.376147
PSP                             4.725247
QAI                            77.462118
SPY                           -49.963644
TIP                           -46.895270

In [37]:
w_tan_adj_tips_summary_statistics = calculate_summary_statistics(df_adj[df1.columns[1:]] @ w_t_adj_tips , annual_factor = 12)
w_tan_adj_tips_summary_statistics.index = ['Tangency Weights with TIPS adjusted']
w_tan_adj_tips_summary_statistics

Mean       Vol    Sharpe       Min  \
Tangency Weights with TIPS adjusted -6.174021  3.655768 -1.688844 -3.892334   

                                          Max  Skewness  Excess Kurtosis  \
Tangency Weights with TIPS adjusted  2.336223 -0.343166          0.71521   

                                     VaR (0.05)  CVaR (0.05)  
Tangency Weights with TIPS adjusted   -2.438227    -2.919097

In [38]:
pd.concat([w_tan_summary_statistics,w_tan_wo_tips_summary_statistics,w_tan_adj_tips_summary_statistics])

Mean       Vol    Sharpe       Min  \
Tangent Weights                     -0.527276  0.328513 -1.605037 -0.293516   
Tangency Weights with TIPS dropped  -0.580806  0.362125 -1.603881 -0.331660   
Tangency Weights with TIPS adjusted -6.174021  3.655768 -1.688844 -3.892334   

                                          Max  Skewness  Excess Kurtosis  \
Tangent Weights                      0.243644 -0.051780         0.105926   
Tangency Weights with TIPS dropped   0.266782 -0.085016         0.162678   
Tangency Weights with TIPS adjusted  2.336223 -0.343166         0.715210   

                                     VaR (0.05)  CVaR (0.05)  
Tangent Weights                       -0.198018    -0.244705  
Tangency Weights with TIPS dropped    -0.220747    -0.272286  
Tangency Weights with TIPS adjusted   -2.438227    -2.919097

Based on the analysis, excluding TIPS from the bond space has a negligible impact on the portfolio's performance, but the mean and vol of Tangency Weights will be significantly impacted if TIPS is adjusted to be 0.0012 higher. Considering TIPS' moderate performance and low correlation with other bonds, I believe it should still be considered as a separate asset. 

3. Allocations

In [39]:
# set the targeted mean excess return
target_mean = 0.01
w_t_without_tips.loc['TIP'] = 0
wts = pd.DataFrame(index = df3.columns, columns = ['tangency','tangency with TIPS dropped',
                                                              'tangency with TIPS adjusted','equal weights',
                                                              'risk parity','regularized'])
wts.loc[:,'tangency'] = tangency_weights(df3, cov_mat = 1).values
wts.loc[:,'tangency with TIPS dropped'] = w_t_without_tips.values
wts.loc[:,'tangency with TIPS adjusted'] = w_t_adj_tips.values
wts.loc[:,'equal weights'] = 1/len(df3.columns)
wts.loc[:,'risk parity'] = 1/df3.var()
wts.loc[:,'regularized'] = tangency_weights(df3, cov_mat = 0.5).values

wts1 = wts*target_mean / (df3.mean()@wts)
wtsADJ = wts*target_mean / (df_adj.mean()@wts)
x = wtsADJ.iloc[:, 2]
wts = pd.concat([wts1.iloc[:, :2], x, wts1.iloc[:, 3:]], axis=1)
wts

C:\Users\Yoga260\AppData\Local\Temp\ipykernel_11732\1583186572.py:8: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  wts.loc[:,'tangency'] = tangency_weights(df3, cov_mat = 1).values
C:\Users\Yoga260\AppData\Local\Temp\ipykernel_11732\1583186572.py:9: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  wts.loc[:,'tangency with TIPS dropped'] = w_t_without_tips.values
C:\Users\Yoga260\AppData\Local\Temp\ipykernel_11732\1583186572.py:10: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead 

tangency  tangency with TIPS dropped  tangency with TIPS adjusted  \
BWX -0.621364                   -0.623116                    -0.568707   
DBC  0.025317                    0.018795                    -0.034434   
EEM  0.140124                    0.137236                     0.099573   
EFA -0.120777                   -0.111694                    -0.028211   
HYG  0.189482                    0.180089                     0.086688   
IEF  1.065661                    0.998417                     0.361604   
IYR -0.199617                   -0.204576                    -0.221110   
PSP -0.063041                   -0.067127                    -0.091841   
QAI -1.643297                   -1.648106                    -1.505575   
SPY  1.116095                    1.113473                     0.971107   
TIP -0.116167                   -0.000000                     0.911470   

     equal weights  risk parity  regularized  
BWX       0.271575     0.522879    -0.675374  
DBC       0.271575     0.122905    -0.136195  
EEM       0.271575     0.107717    -0.066165  
EFA       0.271575     0.150554     0.037302  
HYG       0.271575     0.583857     0.316781  
IEF       0.271575     0.873276     0.393506  
IYR       0.271575     0.121273     0.080277  
PSP       0.271575     0.075284     0.049547  
QAI       0.271575     1.452202    -0.156500  
SPY       0.271575     0.170397     0.623712  
TIP       0.271575     1.327461     0.322429

In [40]:
summary_statistics = calculate_summary_statistics(df3 @ wts, annual_factor=12)
summary1 = calculate_summary_statistics(df_adj @ wts, annual_factor=12)
tangency_row = summary1.loc["tangency with TIPS adjusted"]
new_df = pd.DataFrame()
new_df = pd.concat([summary_statistics.iloc[:2], tangency_row.to_frame().T, summary_statistics.iloc[3:]], ignore_index=False)
new_df

Mean       Vol    Sharpe       Min       Max  \
tangency                     0.12  0.074765  1.605037 -0.055450  0.066800   
tangency with TIPS dropped   0.12  0.074818  1.603881 -0.055120  0.068524   
tangency with TIPS adjusted  0.12  0.071055  1.688844 -0.045407  0.075652   
equal weights                0.12  0.291527  0.411625 -0.349896  0.232113   
risk parity                  0.12  0.309515  0.387703 -0.322150  0.220761   
regularized                  0.12  0.102604  1.169541 -0.070349  0.106592   

                             Skewness  Excess Kurtosis  VaR (0.05)  \
tangency                     0.051780         0.105926   -0.025996   
tangency with TIPS dropped   0.085016         0.162678   -0.025474   
tangency with TIPS adjusted  0.343166         0.715210   -0.021398   
equal weights               -0.616344         2.157157   -0.121427   
risk parity                 -0.590209         1.565413   -0.140760   
regularized                 -0.094926         1.024310   -0.040381   

                             CVaR (0.05)  
tangency                       -0.033186  
tangency with TIPS dropped     -0.032835  
tangency with TIPS adjusted    -0.029761  
equal weights                  -0.189085  
risk parity                    -0.203008  
regularized                    -0.056345

By comparing the results of the four allocations, we found that the "tangency" allocation achieves the hoghest Sharpe ratio, followed by "regularized", then "equal weights" and "risk parity" being much lower. This indicates that, after constraining the minimum return, we can find the tangency weights on the efficient frontier, achieving the maximum Sharpe ratio. 